src/Amalgam/ThreadPool.cpp

```diff
@@ -1,110 +1,143 @@
 
-	//no active threads in the pool to start
-	numActiveThreads = 0;
-	
+	maxNumActiveThreads = 1;
+	numActiveThreads = 1;
+	numReservedThreads = 0;
+	numThreadsToTransitionToReserved = 0;
+
+	SetMaxNumActiveThreads(max_num_active_threads);
+
 	mainThreadId = std::this_thread::get_id();
 }
 
-void ThreadPool::ChangeThreadPoolSize(size_t new_max_num_threads)
+void ThreadPool::SetMaxNumActiveThreads(int32_t new_max_num_active_threads)
 {
-	std::unique_lock<std::mutex> threads_lock(threadsMutex);
-	std::unique_lock<std::mutex> queue_lock(taskQueueMutex);
+	std::unique_lock<std::mutex> lock(threadsMutex);
 
 	//don't need to change anything
-	if(new_max_num_threads == threads.size())
+	if(new_max_num_active_threads == maxNumActiveThreads || new_max_num_active_threads < 1)
 		return;
 
 	//if reducing thread count, clean up all jobs and clear out all threads
-	if(new_max_num_threads < threads.size())
+	if(new_max_num_active_threads < maxNumActiveThreads)
 	{
+		lock.unlock();
 		ShutdownAllThreads();
+		lock.lock();
+
 		threads.clear();
 
 		//no longer shutting down, allow to build up threads
 		shutdownThreads = false;
-	}
 
-	size_t num_new_threads = new_max_num_threads - threads.size();
+		//reset other stats
+		maxNumActiveThreads = 1;
+		numActiveThreads = 1;
+		numReservedThreads = 0;
+	}
 
 	//place an empty idle task for each thread waiting for work
-	for(size_t i = 0; i < num_new_threads; i++)
-	{
-		threads.emplace_back(
-			[this]
+	//but current thread counts as one
+	for(int32_t i = static_cast<int32_t>(threads.size()); i < new_max_num_active_threads - 1; i++)
+		AddNewThread();
+
+	maxNumActiveThreads = new_max_num_active_threads;
+
+	//notify all just in case a new task was added as the threads were being created
+	// but unlock to allow threads to proceed
+	lock.unlock();
+	waitForTask.notify_all();
+}
+
+void ThreadPool::AddNewThread()
+{
+	threads.emplace_back(
+		[this]
 		{
 			//count this thread as active during startup
 			//this is important, as the inner loop assumes the default state of the thread is to count itself
 			//so the number of threads doesn't change when switching between a completed task and a new one
 			numActiveThreads++;
 
+			std::unique_lock<std::mutex> lock(threadsMutex);
+
 			//infinite loop waiting for work
 			for(;;)
 			{
-				//container for the task
-				std::function<void()> task;
-
-				//fetch from queue
+				if(numThreadsToTransitionToReserved > 0)
+				{
+					//go into reserved
+					numActiveThreads--;
+					numThreadsToTransitionToReserved--;
+					numReservedThreads++;
+
+					//wait until either shutting down or a thread is requested to come out of reserved
+					waitForActivate.wait(lock,
+						[this] { return numThreadsToTransitionToReserved < 0 || shutdownThreads; });
+
+					//only can make it here if shutting down (otherwise taskQueue has something in it)
+					if(shutdownThreads)
+						return;
+
+					//coming out of reserved, go active unless no task
```

Comment:
Is this comment accurate? Not sure about the "... unless no task" part is relevant here.